### ISSUE TITLE
Multiple StrictMode warnings during app initialization

### ISSUE THREAD

**Issue by gmerinojimenez** [Swrve/swrve-android-sdk#285]
I'm testing the SDK integration in our app and I'm noticing a lot of strict mode warnings during the startup of the application (11 in my latest tests).

Checking the code, looks like that during the init() method, executed in the onActivityCreated callback method, which runs on the UI thread, there are multiple accesses to disk (initialize the cache directory, open and read from the database...). Are you aware of this? According to the warnings the performance penalty seems noticeable. 

My suggestion is, as google says, move the IO work to a background job.

A few examples:
`04-10 19:10:35.697 19621 19621 D StrictMode: StrictMode policy violation; ~duration=236 ms: android.os.strictmode.DiskReadViolation
04-10 19:10:35.697 19621 19621 D StrictMode: 	at android.os.StrictMode$AndroidBlockGuardPolicy.onReadFromDisk(StrictMode.java:1504)
04-10 19:10:35.697 19621 19621 D StrictMode: 	at java.io.UnixFileSystem.checkAccess(UnixFileSystem.java:251)
04-10 19:10:35.697 19621 19621 D StrictMode: 	at java.io.File.exists(File.java:1168)
04-10 19:10:35.697 19621 19621 D StrictMode: 	at android.app.ContextImpl.getDataDir(ContextImpl.java:2301)
04-10 19:10:35.697 19621 19621 D StrictMode: 	at android.app.ContextImpl.getCacheDir(ContextImpl.java:695)
04-10 19:10:35.697 19621 19621 D StrictMode: 	at android.content.ContextWrapper.getCacheDir(ContextWrapper.java:269)
04-10 19:10:35.697 19621 19621 D StrictMode: 	at com.swrve.sdk.SwrveBase.initCacheFolder(SwrveBase.java:240)
04-10 19:10:35.697 19621 19621 D StrictMode: 	at com.swrve.sdk.SwrveBase.init(SwrveBase.java:131)
04-10 19:10:35.697 19621 19621 D StrictMode: 	at com.swrve.sdk.SwrveBase.onCreate(SwrveBase.java:103)
04-10 19:10:35.697 19621 19621 D StrictMode: 	at com.swrve.sdk.SwrveBase.onActivityCreated(SwrveBase.java:1848)
04-10 19:10:35.697 19621 19621 D StrictMode: 	at android.app.Application.dispatchActivityCreated(Application.java:248)
04-10 19:10:35.697 19621 19621 D StrictMode: 	at android.app.Activity.onCreate(Activity.java:1048)
04-10 19:10:35.697 19621 19621 D StrictMode: 	at android.support.v4.app.SupportActivity.onCreate(SupportActivity.java:66)
04-10 19:10:35.697 19621 19621 D StrictMode: 	at android.support.v4.app.FragmentActivity.onCreate(FragmentActivity.java:285)
04-10 19:10:35.697 19621 19621 D StrictMode: 	at android.support.v7.app.AppCompatActivity.onCreate(AppCompatActivity.java:84)`

`04-10 19:10:35.700 19621 19621 D StrictMode: StrictMode policy violation; ~duration=235 ms: android.os.strictmode.DiskReadViolation
04-10 19:10:35.700 19621 19621 D StrictMode: 	at android.os.StrictMode$AndroidBlockGuardPolicy.onReadFromDisk(StrictMode.java:1504)
04-10 19:10:35.700 19621 19621 D StrictMode: 	at android.database.sqlite.SQLiteConnection.applyBlockGuardPolicy(SQLiteConnection.java:1055)
04-10 19:10:35.700 19621 19621 D StrictMode: 	at android.database.sqlite.SQLiteConnection.executeForLong(SQLiteConnection.java:609)
04-10 19:10:35.700 19621 19621 D StrictMode: 	at android.database.sqlite.SQLiteSession.executeForLong(SQLiteSession.java:652)
04-10 19:10:35.700 19621 19621 D StrictMode: 	at android.database.sqlite.SQLiteStatement.simpleQueryForLong(SQLiteStatement.java:107)
04-10 19:10:35.700 19621 19621 D StrictMode: 	at android.database.DatabaseUtils.longForQuery(DatabaseUtils.java:842)
04-10 19:10:35.700 19621 19621 D StrictMode: 	at android.database.DatabaseUtils.longForQuery(DatabaseUtils.java:830)
04-10 19:10:35.700 19621 19621 D StrictMode: 	at android.database.sqlite.SQLiteDatabase.getPageSize(SQLiteDatabase.java:1043)
04-10 19:10:35.700 19621 19621 D StrictMode: 	at android.database.sqlite.SQLiteDatabase.setMaximumSize(SQLiteDatabase.java:1026)
04-10 19:10:35.700 19621 19621 D StrictMode: 	at com.swrve.sdk.localstorage.SQLiteLocalStorage.<init>(SQLiteLocalStorage.java:50)
04-10 19:10:35.700 19621 19621 D StrictMode: 	at com.swrve.sdk.SwrveBase.openLocalStorageConnection(SwrveBase.java:297)
04-10 19:10:35.700 19621 19621 D StrictMode: 	at com.swrve.sdk.SwrveBase.init(SwrveBase.java:134)
04-10 19:10:35.700 19621 19621 D StrictMode: 	at com.swrve.sdk.SwrveBase.onCreate(SwrveBase.java:103)
04-10 19:10:35.700 19621 19621 D StrictMode: 	at com.swrve.sdk.SwrveBase.onActivityCreated(SwrveBase.java:1848)
04-10 19:10:35.700 19621 19621 D StrictMode: 	at android.app.Application.dispatchActivityCreated(Application.java:248)
04-10 19:10:35.700 19621 19621 D StrictMode: 	at android.app.Activity.onCreate(Activity.java:1048)
04-10 19:10:35.700 19621 19621 D StrictMode: 	at android.support.v4.app.SupportActivity.onCreate(SupportActivity.java:66)
04-10 19:10:35.700 19621 19621 D StrictMode: 	at android.support.v4.app.FragmentActivity.onCreate(FragmentActivity.java:285)
04-10 19:10:35.700 19621 19621 D StrictMode: 	at android.support.v7.app.AppCompatActivity.onCreate(AppCompatActivity.java:84)`

Total time spent during app startup in Swrve Strict mode warnings:
`04-10 17:01:50.760 17887 17887 D StrictMode: StrictMode policy violation; ~duration=230 ms
04-10 17:01:50.761 17887 17887 D StrictMode: StrictMode policy violation; ~duration=230 ms
04-10 17:01:50.762 17887 17887 D StrictMode: StrictMode policy violation; ~duration=229 ms
04-10 17:01:50.763 17887 17887 D StrictMode: StrictMode policy violation; ~duration=229 ms
04-10 17:01:50.764 17887 17887 D StrictMode: StrictMode policy violation; ~duration=228 ms
04-10 17:01:50.765 17887 17887 D StrictMode: StrictMode policy violation; ~duration=226 ms
04-10 17:01:50.766 17887 17887 D StrictMode: StrictMode policy violation; ~duration=225 ms
04-10 17:01:50.767 17887 17887 D StrictMode: StrictMode policy violation; ~duration=224 ms
04-10 17:01:50.768 17887 17887 D StrictMode: StrictMode policy violation; ~duration=223 ms
04-10 17:01:51.033 17887 17887 D StrictMode: StrictMode policy violation; ~duration=95 ms:
04-10 17:01:51.033 17887 17887 D StrictMode: StrictMode policy violation; ~duration=94 ms:`

**Comment by dominicmarmionswrve** [comment]
Hi @gmerinojimenez, thank you for reporting the issue. We'll investigate this and come back to you with more information soon.

**Comment by Dimezis** [comment]
Any updates on this?
I also noticed that initialization in `SwrveBase.onActivityCreated` is extremely slow

**Comment by Dimezis** [comment]
Can we move `SwrveBase.init` to a worker thread?
[It blocks the main thread for almost 400ms on some devices](http://prntscr.com/oczljg)
`SwrveSDK.event` also takes quite a lot of time, as you see

**Comment by davidbreenwex** [comment]
Hi @Dimezis 

This is something we are looking at for our next major SDK update. (v7.0)
Unfortunately this is not expected in the immediate future though.

I'll update you when work on this has been scheduled. 

Thanks, 
David

**Comment by nimeacuerdo** [comment]
Hi, any news on this? `6.3.0` still has these problems.

**Comment by davidbreenwex** [comment]
Hi @nimeacuerdo,

This work to be included in the next major version release 7.0. 
This is currently scheduled for the end of the next quarter. I'll update you once the fix goes out. 

Thanks, 
David

**Comment by gmerinojimenez** [comment]
Hi @davidbreenwex.
We're testing 7.0 and the warnings remain.

Which is expected considering you're opening a database in main thread during activity creation:

```                         D      at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:298)
                         D      at com.swrve.sdk.localstorage.SQLiteLocalStorage.<init>(SourceFile:53)
                         D      at com.swrve.sdk.SwrveBase.openLocalStorageConnection(SourceFile:320)
                         D      at com.swrve.sdk.SwrveBase.init(SourceFile:136)
                         D      at com.swrve.sdk.SwrveBase.onCreate(SourceFile:104)
                         D      at com.swrve.sdk.SwrveBase.onActivityCreated(SourceFile:1943)```

That is taking 248ms in an Oneplus 5T. My question is: why don't you use another thread to initialize your SDK?

**Comment by davidbreenwex** [comment]
Hi @gmerinojimenez ,

Unfortunately these changes didn't make it into the 7.0 release. The project was scoped and it looks like it will take more effort than we initially thought. This is in the planning phase now. I'll update you once it has been scheduled. 

Best, 
David

**Comment by gmerinojimenez** [comment]
I've created a Pull Request to fix this since it's a huge performance penalty in our app.

Could you please take a look at it?
https://github.com/Swrve/swrve-android-sdk/pull/292

**Comment by nimeacuerdo** [comment]
Is this ever gonna be addressed?

Thanks in advance!

**Comment by indieboyalex** [comment]
Is there any news on this?  We are seeing a LOT of ANRs in Google Vitals, all of which link back to the Swrve SDK

**Type "main" prio=5 tid=1 Blocked**

```
  at com.swrve.sdk.localstorage.SwrveMultiLayerLocalStorage.getCombinedFirstNEvents (SwrveMultiLayerLocalStorage.java:108)
  at com.swrve.sdk.localstorage.SwrveMultiLayerLocalStorage.hasQueuedEvents (SwrveMultiLayerLocalStorage.java:103)
  at com.swrve.sdk.SwrveBase._sendQueuedEvents (SwrveBase.java:552)
  at com.swrve.sdk.SwrveBase.sendQueuedEvents (SwrveBase.java:1345)
  at com.swrve.sdk.SwrveBase._onResume (SwrveBase.java:615)
  at com.swrve.sdk.SwrveBase.onResume (SwrveBase.java:1380)
  at com.swrve.sdk.SwrveBase.onActivityResumed (SwrveBase.java:1976)
  at android.app.Application.dispatchActivityResumed (Application.java:251)
  at android.app.Activity.onResume (Activity.java:1424)
  at androidx.fragment.app.FragmentActivity.onResume (FragmentActivity.java:455)
```

**Comment by Sergio-Mira** [comment]
@iNdieboyjeff that specific issue got fixed on 7.3.2.
In terms of resolving all strict mode issues it is a bigger job requiring a redesign of the SDK to make almost all calls async.

**Comment by dominicmarmionswrve** [comment]
Apologies for the long delay. This got fixed in `release-8_0_0`. 🎉 

I'll close this issue now but if you see any further issues please reopen or create a new issue. Thanks for bringing it to our attention!
